### PR TITLE
pubsub: add a test that can be used to characterize Receive performance

### PR DESCRIPTION
**Pull request by vangent**
Sample output with the default settings:

```
$ go test --run TestReceivePerformance
Elapsed (sec)	Msgs/sec	Max Messages
0.000081	12318.001527	1
0.000223	7067487.437541	1000
1.000242	751985.161829	1000
2.000659	792669.396619	1000
3.001288	810490.673882	1000
4.001931	803483.069525	1000
5.002907	787231.494381	1000
6.003415	735626.551089	1000
7.003970	801555.034357	1000
8.005067	805116.952557	1000
9.005451	701730.566339	1000
PASS
ok  	gocloud.dev/pubsub	10.097s
```

Updates #1441.

My plan is to use this to experiment with different settings in pubsub.go. For example:

* What happens if `ReceiveBatch` takes 250ms every time?
* What happens if `ReceiveBatch` is fast but returns fewer than `maxMessages`?
* What happens if the application never calls `Ack`?
* What happens if message processing is very slow?
* What happens if message processing usually fast but occasionally slow?

### CODE REVIEW

**Comment by vangent** [comment]
I'm probably going to end up turning this test into a suite of profiles; the test will run all the profiles and the output will be extended to include a column with the profile name.

Then I'll make a Sheet with 2 sheets, so you can easily do an A/B test by importing the output with A into sheet 1, output with B into sheet 2, and see graphs comparing them for each profile.

File: pubsub/benchmark_test.go [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+// TestReceivePerformance enables characterization of Receive under various
+// situations. Tune the const and var parameters above, disable the Skip,
+// and the test will output a series of timepoints with throughput and the
+// batch size being sent to ReceiveBatch.
+func TestReceivePerformance(t *testing.T) {
+	t.Skip("Skipping by default")
+
+	msgs := make([]*driver.Message, 1000)
+	for i := 0; i < 1000; i++ {
```
Comment (jba) [inline pubsub/benchmark_test.go]: for i := range msgs

File: pubsub/benchmark_test.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+	for i := 0; i < nGoRoutines; i++ {
+		grp.Go(func() error {
+			// Each goroutine loops until ctx is canceled.
+			for {
+				m, err := sub.Receive(ctx)
+				if err == context.Canceled {
+					return nil
+				}
+				// TODO(rvangent): This is annoying. Maybe retry should return
+				// context.Canceled instead of wrapping if it never retried?
```
Comment (jba) [inline pubsub/benchmark_test.go]: Then wouldn't callers of retry have to check for both kinds of errors?

File: pubsub/benchmark_test.go (hunk 1 above)
Comment (vangent) [reply pubsub/benchmark_test.go]: Done.

File: pubsub/benchmark_test.go (hunk 2 above)
Comment (vangent) [reply pubsub/benchmark_test.go]: Filed #1591 to discuss.

**Pull request merged**